projectq/cengines/_replacer/_decompositionhandling.py

```diff
@@ -11,67 +11,101 @@
 #   limitations under the License.
 
 from ._decomposition import Decomposition
+from projectq.ops import BasicGate
 
-decompositions = dict()
 
-
-class ThisIsNotAGateClassError(Exception):
+class ThisIsNotAGateClassError(TypeError):
     pass
 
 
-def register_decomposition(gate_class, gate_decomposer, gate_recognizer=None):
+class DecompositionRule:
```

Comment:
Well, DecompositionRule requires a gate class for indexing purposes but there's currently no way to compute the inverse class when construction the inverse DecompositionRule.

I'm not too worried about it. `_Decomposition` is a private class, and I generally go by the rule that anything you do in the privacy of your own module is fine by me. It's easy to change stuff that isn't coupled.